tag_tracer.go

```diff
@@ -0,0 +1,298 @@
+	tag := topicTag(topic)
+	t.cmgr.TagPeer(p, tag, GossipSubConnTagValueMeshPeer)
+}
+
+func (t *tagTracer) untagMeshPeer(p peer.ID, topic string) {
+	tag := topicTag(topic)
+	t.cmgr.UntagPeer(p, tag)
+}
+
+func topicTag(topic string) string {
+	return fmt.Sprintf("pubsub:%s", topic)
+}
+
+func (t *tagTracer) addDeliveryTag(topic string) {
+	if t.decayer == nil {
+		return
+	}
+
+	t.Lock()
+	defer t.Unlock()
+	tag, err := t.decayingDeliveryTag(topic)
+	if err != nil {
+		log.Warnf("unable to create decaying delivery tag: %s", err)
+		return
+	}
+	t.decaying[topic] = tag
+}
+
+func (t *tagTracer) removeDeliveryTag(topic string) {
+	t.Lock()
+	defer t.Unlock()
+	delete(t.decaying, topic)
+}
+
+func (t *tagTracer) decayingDeliveryTag(topic string) (connmgr.DecayingTag, error) {
+	if t.decayer == nil {
+		return nil, fmt.Errorf("connection manager does not support decaying tags")
+	}
+	name := fmt.Sprintf("pubsub-deliveries:%s", topic)
+
+	// decrement tag value by GossipSubConnTagDecayAmount at each decay interval
+	decayFn := func(value connmgr.DecayingValue) (after int, rm bool) {
+		v := value.Value - GossipSubConnTagDecayAmount
+		return v, v <= 0
+	}
+
+	// bump up to max of GossipSubConnTagMessageDeliveryCap
+	bumpFn := func(value connmgr.DecayingValue, delta int) (after int) {
+		val := value.Value + delta
+		if val > GossipSubConnTagMessageDeliveryCap {
+			return GossipSubConnTagMessageDeliveryCap
+		}
+		return val
+	}
+
+	return t.decayer.RegisterDecayingTag(name, GossipSubConnTagDecayInterval, decayFn, bumpFn)
+}
+
+func (t *tagTracer) bumpDeliveryTag(p peer.ID, topic string) error {
+	t.RLock()
+	defer t.RUnlock()
+
+	tag, ok := t.decaying[topic]
+	if !ok {
+		return fmt.Errorf("no decaying tag registered for topic %s", topic)
+	}
+	return tag.Bump(p, GossipSubConnTagBumpMessageDelivery)
+}
+
+func (t *tagTracer) bumpTagsForMessage(p peer.ID, msg *Message) {
+	for _, topic := range msg.TopicIDs {
+		err := t.bumpDeliveryTag(p, topic)
+		if err != nil {
+			log.Warnf("error bumping delivery tag: %s", err)
+		}
+	}
+}
+
+// nearFirstPeers returns the peers who delivered the message while it was still validating
+func (t *tagTracer) nearFirstPeers(msg *Message) []peer.ID {
+	t.Lock()
+	defer t.Unlock()
+	drec := t.deliveries.getRecord(t.msgID(msg.Message))
+	nearFirstPeers := make([]peer.ID, 0, len(drec.peers))
+	// defensive check that this is the first delivery trace -- delivery status should be unknown
+	if drec.status != deliveryUnknown {
+		log.Warnf("unexpected delivery trace: message from %s was first seen %s ago and has delivery status %d", msg.ReceivedFrom, time.Now().Sub(drec.firstSeen), drec.status)
+		return nearFirstPeers
+	}
+
+	drec.status = deliveryValid
+	drec.validated = time.Now()
+
+	for p := range drec.peers {
+		// this check is to make sure a peer can't send us a message twice and get a double count
+		// if it is a first delivery.
+		if p != msg.ReceivedFrom {
+			nearFirstPeers = append(nearFirstPeers, p)
+		}
+	}
```

Comment:
We might want to nil the `drec.peers` map before we return the near first peers, so as to avoid holding onto memory unnecessarily. It shouldn't be used again after the delivery status has been marked as `deliveryValid`.